test/new-e2e/tests/agent-shared-components/hostname/imdsv2_transition_common_test.go

```diff
@@ -0,0 +1,195 @@
+	"github.com/DataDog/datadog-agent/test/new-e2e/pkg/utils/e2e/client/agentclient"
+)
+
+type Meta struct {
+	Hostname                 string `json:"hostname"`
+	LegacyResolutionHostname string `json:"legacy-resolution-hostname,omitempty"`
+}
+
+type baseHostnameSuite struct {
+	e2e.BaseSuite[environments.Host]
+
+	hostnameMetadata Meta
+}
+
+func TestBaseHostnameSuite(t *testing.T) {
+	e2e.Run(t, &baseHostnameSuite{}, e2e.WithProvisioner(awshost.ProvisionerNoFakeIntake()))
+}
+
+func requestAgentHostnameMetadataPayload(v *baseHostnameSuite) {
+	v.T().Helper()
+
+	// Get the hostname and legacy-resolution-hostname from the agent status
+	statusArgs := agentclient.WithArgs([]string{"header", "--json"})
+	statusPayload := v.Env().Agent.Client.Status(statusArgs)
+
+	var data map[string]interface{}
+	if err := json.Unmarshal([]byte(statusPayload.Content), &data); err != nil {
+		v.T().Fatal(err)
+	}
+
+	metaData, err := json.Marshal(data["metadata"].(map[string]interface{})["meta"])
+	if err != nil {
+		v.T().Fatal(err)
+	}
+
+	var hostnameMeta Meta
+	if err := json.Unmarshal(metaData, &hostnameMeta); err != nil {
+		v.T().Fatal(err)
+	}
+	v.hostnameMetadata = hostnameMeta
+}
+
+// retrieveMetadata retrieves the instance ID and OS hostname from the EC2 metadata
+func retrieveMetadata(v *baseHostnameSuite) (string, string) {
+	metadata := client.NewEC2Metadata(v.T(), v.Env().RemoteHost.Host, v.Env().RemoteHost.OSFamily)
+	instanceID := metadata.Get("instance-id")
+	osHostname := strings.Split(metadata.Get("hostname"), ".")[0]
+	return instanceID, osHostname
+}
+
+// runHostnameTest runs the hostname test with the given parameters
+func runHostnameTest(v *baseHostnameSuite, instanceOpts []awshost.ProvisionerOption, imdsV1Enable bool, tt struct {
+	name                     string
+	ec2PreferIMDSv2          bool
+	legacyResolutionHostname bool
+}) {
+	v.T().Run(tt.name, func(t *testing.T) {
+		v.UpdateEnv(awshost.ProvisionerNoFakeIntake(instanceOpts...))
+
+		requestAgentHostnameMetadataPayload(v)
+
+		instanceID, osHostname := retrieveMetadata(v)
+
+		assert.NotEmpty(t, v.hostnameMetadata.Hostname)
+
+		// legacy-resolution-hostname should only be set when IMDSv1 is disabled, ec2_prefer_imdsv2 is disabled and legacy-resolution-hostname is enabled
+		if tt.legacyResolutionHostname && !imdsV1Enable && !tt.ec2PreferIMDSv2 {
+			assert.NotEmpty(t, v.hostnameMetadata.LegacyResolutionHostname)
+			assert.NotEqual(t, v.hostnameMetadata.Hostname, v.hostnameMetadata.LegacyResolutionHostname)
+			assert.Equal(t, v.hostnameMetadata.LegacyResolutionHostname, osHostname)
+		} else {
+			assert.Empty(t, v.hostnameMetadata.LegacyResolutionHostname)
+		}
+
+		// The hostname should be the instance ID when IMDSv1 or IMDSv2 is enabled
+		if imdsV1Enable || tt.ec2PreferIMDSv2 || tt.legacyResolutionHostname {
+			assert.Equal(t, instanceID, v.hostnameMetadata.Hostname)
+		} else {
+			assert.Equal(t, osHostname, v.hostnameMetadata.Hostname)
+		}
```

Comment:
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**🥜 nitpick:** Since you test the same variable in both cases, IMO it is clearer to refactor it this way: ‏
```suggestion
		expectedHostname  := osHostname
		
		// The hostname should be the instance ID when IMDSv1 or IMDSv2 is enabled
		if imdsV1Enable || tt.ec2PreferIMDSv2 || tt.legacyResolutionHostname {
			expectedHostname = instanceID
		}
		
		assert.Equal(t, expectedHostname, v.hostnameMetadata.Hostname)
```

Reply:
good suggestion I'll just revert the boolean logic since most of the time we are awaiting the instanceID